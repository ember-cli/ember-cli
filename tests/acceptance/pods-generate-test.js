'use strict';

const ember = require('../helpers/ember');
const replaceFile = require('ember-cli-internal-test-helpers/lib/helpers/file-utils').replaceFile;
const fs = require('fs-extra');
const path = require('path');
let root = process.cwd();
let tmproot = path.join(root, 'tmp');
const mkTmpDirIn = require('../../lib/utilities/mk-tmp-dir-in');

const Blueprint = require('../../lib/models/blueprint');
const BlueprintNpmTask = require('ember-cli-internal-test-helpers/lib/helpers/disable-npm-on-blueprint');

const chai = require('../chai');
let expect = chai.expect;
let file = chai.file;

describe('Acceptance: ember generate pod', function () {
  this.timeout(60000);

  let tmpdir;

  before(function () {
    BlueprintNpmTask.disableNPM(Blueprint);
  });

  after(function () {
    BlueprintNpmTask.restoreNPM(Blueprint);
  });

  beforeEach(async function () {
    tmpdir = await mkTmpDirIn(tmproot);
    process.chdir(tmpdir);
  });

  afterEach(function () {
    process.chdir(root);
    return fs.remove(tmproot);
  });

  function initApp() {
    return ember(['init', '--name=my-app', '--skip-npm', '--skip-bower']).then(addJSHint);
  }

  function addJSHint() {
    let pkg = fs.readJsonSync('package.json');
    pkg.devDependencies['ember-cli-jshint'] = '*';
    fs.writeJsonSync('package.json', pkg);
  }

  function generate(args) {
    let generateArgs = ['generate'].concat(args);

    return initApp().then(function () {
      return ember(generateArgs);
    });
  }

  function generateWithPrefix(args) {
    let generateArgs = ['generate'].concat(args);

    return initApp().then(function () {
      replaceFile('config/environment.js', '(var|let|const) ENV = {', "$1 ENV = {\npodModulePrefix: 'app/pods', \n");
      return ember(generateArgs);
    });
  }

  it('blueprint foo --pod', async function () {
    await generate(['blueprint', 'foo', '--pod']);

    expect(file('blueprints/foo/index.js').content).to.matchSnapshot();
  });

  it('blueprint foo/bar --pod', async function () {
    await generate(['blueprint', 'foo/bar', '--pod']);

    expect(file('blueprints/foo/bar/index.js').content).to.matchSnapshot();
  });

  it('http-mock foo --pod', async function () {
    await generate(['http-mock', 'foo', '--pod']);

    expect(file('server/index.js')).to.contain('mocks.forEach(route => route(app));');

    expect(file('server/mocks/foo.js').content).to.matchSnapshot();

    expect(file('server/.jshintrc')).to.contain('{\n  "node": true\n}');
  });

  it('http-mock foo-bar --pod', async function () {
    await generate(['http-mock', 'foo-bar', '--pod']);

    expect(file('server/index.js')).to.contain('mocks.forEach(route => route(app));');

    expect(file('server/mocks/foo-bar.js').content).to.matchSnapshot();

    expect(file('server/.jshintrc')).to.contain('{\n  "node": true\n}');
  });

  it('http-proxy foo --pod', async function () {
    await generate(['http-proxy', 'foo', 'http://localhost:5000', '--pod']);

    expect(file('server/index.js')).to.contain('proxies.forEach(route => route(app));');

    expect(file('server/proxies/foo.js').content).to.matchSnapshot();

    expect(file('server/.jshintrc')).to.contain('{\n  "node": true\n}');
  });

  it('uses blueprints from the project directory', async function () {
    await initApp();
    await fs.outputFile(
      'blueprints/foo/files/app/foos/__name__.js',
      "import Ember from 'ember';\n" + 'export default Ember.Object.extend({ foo: true });\n'
    );

    await ember(['generate', 'foo', 'bar', '--pod']);

    expect(file('app/foos/bar.js')).to.contain('foo: true');
  });

  it('allows custom blueprints to override built-ins', async function () {
    await initApp();
    await fs.outputFile(
      'blueprints/controller/files/app/__path__/__name__.js',
      "import Ember from 'ember';\n\n" + 'export default Ember.Controller.extend({ custom: true });\n'
    );

    await ember(['generate', 'controller', 'foo', '--pod']);

    expect(file('app/foo/controller.js')).to.contain('custom: true');
  });

  it('passes custom cli arguments to blueprint options', async function () {
    await initApp();
    await fs.outputFile(
      'blueprints/customblue/files/app/__name__.js',
      'Q: Can I has custom command? A: <%= hasCustomCommand %>'
    );

    await fs.outputFile(
      'blueprints/customblue/index.js',
      'module.exports = {\n' +
        '  fileMapTokens(options) {\n' +
        '    return {\n' +
        '      __name__(options) {\n' +
        '         return options.dasherizedModuleName;\n' +
        '      }\n' +
        '    };\n' +
        '  },\n' +
        '  locals(options) {\n' +
        '    var loc = {};\n' +
        "    loc.hasCustomCommand = (options.customCommand) ? 'Yes!' : 'No. :C';\n" +
        '    return loc;\n' +
        '  },\n' +
        '};\n'
    );

    await ember(['generate', 'customblue', 'foo', '--custom-command', '--pod']);

    expect(file('app/foo.js')).to.contain('A: Yes!');
  });

  it('correctly identifies the root of the project', async function () {
    await initApp();
    await fs.outputFile(
      'blueprints/controller/files/app/__path__/__name__.js',
      "import Ember from 'ember';\n\n" + 'export default Ember.Controller.extend({ custom: true });\n'
    );

    process.chdir(path.join(tmpdir, 'app'));
    await ember(['generate', 'controller', 'foo', '--pod']);

    process.chdir(tmpdir);
    expect(file('app/foo/controller.js')).to.contain('custom: true');
  });

  // Skip until podModulePrefix is deprecated
  it.skip('podModulePrefix deprecation warning', async function () {
    let result = await generateWithPrefix(['controller', 'foo', '--pod']);

    expect(result.outputStream.join()).to.include(
      '`podModulePrefix` is deprecated and will be' +
        ' removed from future versions of ember-cli. Please move existing pods from' +
        " 'app/pods/' to 'app/'."
    );
  });
});
