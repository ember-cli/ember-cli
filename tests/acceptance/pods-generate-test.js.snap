// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Acceptance: ember generate pod blueprint foo --pod 1`] = `
"'use strict';

module.exports = {
  description: ''

  // locals(options) {
  //   // Return custom template variables here.
  //   return {
  //     foo: options.entity.options.foo
  //   };
  // }

  // afterInstall(options) {
  //   // Perform extra work here.
  // }
};
"
`;

exports[`Acceptance: ember generate pod blueprint foo/bar --pod 1`] = `
"'use strict';

module.exports = {
  description: ''

  // locals(options) {
  //   // Return custom template variables here.
  //   return {
  //     foo: options.entity.options.foo
  //   };
  // }

  // afterInstall(options) {
  //   // Perform extra work here.
  // }
};
"
`;

exports[`Acceptance: ember generate pod http-mock foo --pod 1`] = `
"'use strict';

module.exports = function(app) {
  const express = require('express');
  let fooRouter = express.Router();

  fooRouter.get('/', function(req, res) {
    res.send({
      'foo': []
    });
  });

  fooRouter.post('/', function(req, res) {
    res.status(201).end();
  });

  fooRouter.get('/:id', function(req, res) {
    res.send({
      'foo': {
        id: req.params.id
      }
    });
  });

  fooRouter.put('/:id', function(req, res) {
    res.send({
      'foo': {
        id: req.params.id
      }
    });
  });

  fooRouter.delete('/:id', function(req, res) {
    res.status(204).end();
  });

  // The POST and PUT call will not contain a request body
  // because the body-parser is not included by default.
  // To use req.body, run:

  //    npm install --save-dev body-parser

  // After installing, you need to \`use\` the body-parser for
  // this mock uncommenting the following line:
  //
  //app.use('/api/foo', require('body-parser').json());
  app.use('/api/foo', fooRouter);
};
"
`;

exports[`Acceptance: ember generate pod http-mock foo-bar --pod 1`] = `
"'use strict';

module.exports = function(app) {
  const express = require('express');
  let fooBarRouter = express.Router();

  fooBarRouter.get('/', function(req, res) {
    res.send({
      'foo-bar': []
    });
  });

  fooBarRouter.post('/', function(req, res) {
    res.status(201).end();
  });

  fooBarRouter.get('/:id', function(req, res) {
    res.send({
      'foo-bar': {
        id: req.params.id
      }
    });
  });

  fooBarRouter.put('/:id', function(req, res) {
    res.send({
      'foo-bar': {
        id: req.params.id
      }
    });
  });

  fooBarRouter.delete('/:id', function(req, res) {
    res.status(204).end();
  });

  // The POST and PUT call will not contain a request body
  // because the body-parser is not included by default.
  // To use req.body, run:

  //    npm install --save-dev body-parser

  // After installing, you need to \`use\` the body-parser for
  // this mock uncommenting the following line:
  //
  //app.use('/api/foo-bar', require('body-parser').json());
  app.use('/api/foo-bar', fooBarRouter);
};
"
`;

exports[`Acceptance: ember generate pod http-proxy foo --pod 1`] = `
"'use strict';

const proxyPath = '/foo';

module.exports = function(app) {
  // For options, see:
  // https://github.com/nodejitsu/node-http-proxy
  let proxy = require('http-proxy').createProxyServer({});

  proxy.on('error', function(err, req) {
    console.error(err, req.url);
  });

  app.use(proxyPath, function(req, res, next){
    // include root path in proxied request
    req.url = proxyPath + '/' + req.url;
    proxy.web(req, res, { target: 'http://localhost:5000' });
  });
};
"
`;
